feat: Improve the robustness of SVM

michaelasp opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Improve the error handling of SVM to handle both errors that can skip resources and errors that require a retry. Also skip the reconciliation of resources that do not need migration if they are already versioned past the RV that SVM marked as the beginning of migration.

#### Which issue(s) this PR is related to:
KEP: https://github.com/kubernetes/enhancements/issues/4192
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/4192

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:
Also looking for feedback on other places to improve robustness before moving to beta.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

stlaz commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -271,6 +271,16 @@ func (svmc *SVMController) sync(ctx context.Context, key string) error {
 		if err != nil {
 			return err
 		}
+		objRV, err := convertResourceVersionToInt(accessor.GetResourceVersion())
+		if err != nil {
+			logger.V(4).Error(err, "Unable to migrate resource since its version is not an integer", "namespace", accessor.GetNamespace(), "name", accessor.GetName(), "gvr", gvr.String())
+			return svmc.failMigration(ctx, toBeProcessedSVM, err)
+		}
+
+		if objRV > listResourceVersion {
+			logger.V(6).Info("Resource has a newer resource version than start of migration, does not need migration", "namespace", accessor.GetNamespace(), "name", accessor.GetName(), "gvr", gvr.String())
+			continue
+		}
```
Error handling and RV handling should be at least separate commits

michaelasp replied:
Sure done, also planning on moving to this helper once it merges https://github.com/kubernetes/kubernetes/pull/134330

michaelasp replied:
Kept in this PR but moved to using the helper in places we compare the RV since this is the official helper.

michaelasp committed:
feat: Improve the robustness of SVM

Add handling for retriable errors, terminal cases and unit tests for SVM. Also move the comparisons of resource version to use the built in helper provided by apimachinery.

michaelasp commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -339,6 +346,47 @@ func (svmc *SVMController) sync(ctx context.Context, key string) error {
+	}
+}
+
+func isRetriableError(k8sError error) bool {
```
Anything that prevents the migration from being able to progress, one off the top of my head is that we are migrating CRDs, but the CRD is deleted. That should cause the request to fail, I believe with the `IsInvalid` failure. Those types of errors should be real errors and bubbled up to the migration object.

enj replied:
`utilnet.IsConnectionReset`
`utilnet.IsHTTP2ConnectionLost`

Maybe `EOF`?

michaelasp replied:
Added those to the list of retriable errors. Lmk what you think.

michaelasp commented:
/retest

enj commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -271,6 +270,15 @@ func (svmc *SVMController) sync(ctx context.Context, key string) error {
 		}
+		rvCmp, err := resourceversion.CompareResourceVersion(accessor.GetResourceVersion(), listResourceVersion)
+		if err != nil {
+			logger.V(4).Error(err, "Unable to compare the resource version of the resource", "namespace", accessor.GetNamespace(), "name", accessor.GetName(), "gvr", gvr.String(), "error", err.Error())
+			return err
+		}
+		if rvCmp == 1 {
+			logger.V(6).Info("Resource ignored due to resource version being greater than the SVM checkpoint", "namespace", accessor.GetNamespace(), "name", accessor.GetName(), "gvr", gvr.String())
```
May be helpful to include `accessor.GetResourceVersion()` and `listResourceVersion` in the log messages.

michaelasp replied:
Good point, added!

enj commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -339,6 +347,47 @@ func (svmc *SVMController) sync(ctx context.Context, key string) error {
 	return nil
 }
 
+func isAllowableError(k8sError error) bool {
```
Please revert this part of the change, the only allowable error is conflict.  It is impossible for SSA to return a not found:

```
// set UID so that when a resource gets deleted, we get an "uid mismatch"
// conflict error instead of trying to create it.
```

michaelasp replied:
Done

enj commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -339,6 +336,47 @@ func (svmc *SVMController) sync(ctx context.Context, key string) error {
+	switch {
+	case apierrors.IsServerTimeout(k8sError):
+		return true
+	case apierrors.IsTooManyRequests(k8sError):
+		return true
```
We probably should use this to backoff for the seconds indicated via `SuggestsClientDelay`.

enj commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
Probably should add some unit tests for this...

michaelasp replied:
Added to test the overall sync function. Goes over all the base cases from what I can tell.

enj reviewed (changes requested):
First pass.

enj reviewed (commented):
Have not looked at the unit test yet.

michaelasp commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -162,6 +169,10 @@ func (svmc *SVMController) processNext(ctx context.Context) bool {
 
 	klog.FromContext(ctx).V(2).Info("Error syncing SVM resource, retrying", "svm", key, "err", err)
+	if suggestDelay, ok := apierrors.SuggestsClientDelay(err); ok {
+		svmc.queue.AddAfter(key, time.Second*time.Duration(suggestDelay))
```
Ah yeah good point, especially since we backoff exponentially this would matter. However since it's lowercase it's not exposed to other packages. Is there some way to get this from the svm controller?

michaelasp replied:
I added a new entry to the SVM struct to track the rate limiter, I think this is the best way. LMK what you think

michaelasp commented:
Looks like a flake, but can't find an open issue for it.
/retest

stlaz commented on pkg/controller/storageversionmigrator/storageversionmigrator_test.go:
```diff
@@ -0,0 +1,405 @@
+			},
+			expectErr: false,
+			expectKubeActions: []k8stesting.Action{
+				k8stesting.NewUpdateAction(svmv1alpha1.SchemeGroupVersion.WithResource("storageversionmigrations"), "", nil),
+				k8stesting.NewUpdateAction(svmv1alpha1.SchemeGroupVersion.WithResource("storageversionmigrations"), "", nil),
```
We should test we set the failed condition here

stlaz commented on pkg/controller/storageversionmigrator/storageversionmigrator_test.go:
```diff
@@ -0,0 +1,405 @@
+					newResource("res1", "ns1", "90", "uid1"),
+				}),
+			},
+			dynamicClientPatches: []k8stesting.Action{
```
shouldn't this set `dynamicClientErrors`? Why is it passing now?

michaelasp reviewed (commented):
Updated tests from feedback, added testing to conditions of SVM and caught some misconfigs, thanks @stlaz!

enj commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -339,6 +355,30 @@ func (svmc *SVMController) sync(ctx context.Context, key string) error {
+		apierrors.IsTimeout(k8sError)
+}
+
+func (svmc *SVMController) failMigration(ctx context.Context, toBeProcessedSVM *svmv1alpha1.StorageVersionMigration, err error) error {
```
Can we use this in the `// we can't migrate a resource that doesn't exist in the GC` case above?

michaelasp replied:
Yep, switched to use the helper.

enj commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -235,17 +249,15 @@ func (svmc *SVMController) sync(ctx context.Context, key string) error {
+	rvCmp, err := resourceversion.CompareResourceVersion(gcListResourceVersion, listResourceVersion)
 	if err != nil {
-		return err
+		return fmt.Errorf("error comparing resource versions between GC and SVM resource: %w", err)
```
I think this should fail the migration right?  Retrying will never fix it?

michaelasp replied:
Done! Also added test cases for this.

enj commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
```diff
@@ -264,13 +276,20 @@ func (svmc *SVMController) sync(ctx context.Context, key string) error {
+		rvCmp, err := resourceversion.CompareResourceVersion(accessor.GetResourceVersion(), listResourceVersion)
+		if err != nil {
+			logger.V(4).Error(err, "Unable to compare the resource version of the resource", "namespace", accessor.GetNamespace(), "name", accessor.GetName(), "gvr", gvr.String(), "accessorRV", accessor.GetResourceVersion(), "listResourceVersion", listResourceVersion, "error", err.Error())
+			return err
```
I don't think we can recover from this by retrying so should this also fail the migration?

enj commented on pkg/controller/storageversionmigrator/storageversionmigrator.go:
Also terminal I think?

michaelasp replied:
Yeah while this "should" never happen it is terminal.

enj reviewed (changes requested):
Re-read the controller logic completely to look for the terminal error cases.

enj commented:
LGTM, deferring `/lgtm` to @stlaz and `/approve` to @jpbetz

enj reviewed (changes requested):
On the latest changes.

michaelasp commented:
/cc @janetkuo 
For pkg/controller approval. Thanks!

jpbetz commented:
/approve

enj reviewed (commented):
One comment.

michaelasp commented:
/retest
Unrelated flakes looks like
